src/node_jsvmapi_types.h

```diff
@@ -34,4 +36,48 @@ struct napi_property_descriptor {
   void* data;
 };
 
+enum napi_valuetype {
+  // ES6 types (corresponds to typeof)
+  napi_undefined,
+  napi_null,
+  napi_boolean,
+  napi_number,
+  napi_string,
+  napi_symbol,
+  napi_object,
+  napi_function,
+};
+
+enum napi_typedarray_type {
```

Comment:
Pretty much exactly the surface I was thinking about, yes. Could be a bit lower overhead if the C API exposed tiny wrappers around those v8 IsSomeType() methods. Not sure that overhead is worth worrying about, though.
```c++
napi_status napi_is_number(napi_env e, napi_value v, bool* result) {
  NAPI_PREAMBLE(e);
  CHECK_ARG(result);
  *result = v8impl::V8LocalValueFromJsValue(v)->IsNumber();
  return GET_RETURN_STATUS();
}
```